fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ExpressionAnalyzer.java

```diff
@@ -256,6 +265,37 @@ public Expression visitUnboundAlias(UnboundAlias unboundAlias, ExpressionRewrite
         }
     }
 
+    @Override
+    public Expression visitDereferenceExpression(DereferenceExpression dereferenceExpression,
+            ExpressionRewriteContext context) {
+        Expression expression = dereferenceExpression.child(0).accept(this, context);
+        DataType dataType = expression.getDataType();
+        if (dataType.isStructType()) {
+            StructType structType = (StructType) dataType;
+            StructField field = structType.getField(dereferenceExpression.fieldName);
+            if (field != null) {
+                return new StructElement(expression, dereferenceExpression.child(1));
+            }
+        } else if (dataType.isArrayType()) {
+            if (dereferenceExpression.fieldName.equalsIgnoreCase("size")) {
+                return new Cardinality(expression);
+            }
```

Comment:
so, maybe wo should remove array.size too. because this make dereference has two different behavior on different types